Assets/Plugins/iOS/Shopify/PaymentSession.swift

```diff
@@ -0,0 +1,221 @@
+//
+//  PaymentSession.swift
+//  UnityBuySDK
+//
+//  Created by Shopify.
+//  Copyright © 2017 Shopify Inc. All rights reserved.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import UIKit
+import PassKit
+
+@objc enum PaymentStatus: Int {
+    case Failed
+    case Cancelled
+    
+    func toString() -> String {
+        switch self {
+        case .Cancelled:
+            return "Cancelled"
+        case .Failed:
+            return "Failed"
+        }
+    }
+}
+
+@objc protocol PaymentSessionDelegate : class {
+    func paymentSessionDidFinish(session: PaymentSession,with status: PaymentStatus)
+    func paymentSession(_ session: PaymentSession, didSelect shippingMethod: PKShippingMethod)
+    func paymentSession(_ session: PaymentSession, didSelectShippingContact shippingContact: PKContact)
+    func paymentSession(_ session: PaymentSession, didAuthorize payment: PKPayment)
+}
+
+@objc class PaymentSession: NSObject {
+
+    /// Semaphore that blocks the PKAuthorizationController till summary items
+    /// or shipping methods are updated based on the user's input
+    let updateSemaphore = DispatchSemaphore(value: 0)
```

Comment:
I forgot to mention in the description. But we use a semaphore here so that we can block the before calling the completion closure on `PKPaymentAuthorizationDelegate`. Blocking before calling the completion, allows us to notify a managed function to update the session's summary items and shipping. Managed functions are asynchronous and have no return values so we must wait till we receive another function notifying that all mutations are finished. That is when we unlock the semaphore, and calling the completion closure with the updated summary items.